### PR TITLE
fix: `polycli loadtest` does not stop immediately when one try to interrupt it with `Ctrl+C`

### PR DESCRIPTION
**Pull request by leovct**
# Description

- Move `completeLoadTest` inside `loopFunc` to handle cancellation signals properly.
- Document `runLoadTest` function.

## Jira / Linear Tickets

- [DVT-1069](https://polygon.atlassian.net/browse/DVT-1069)

# Testing

- [x] Make sure `polycli loadtest` stops when performing load test
- [x] Make sure `polycli loadtest` stops when completing load test (summarising)

[DVT-1069]: https://polygon.atlassian.net/browse/DVT-1069?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/loadtest/loadtest.go [hunk 1]
```diff
@@ -347,32 +351,48 @@ func runLoadTest(ctx context.Context) error {
 		overallTimer = new(time.Timer)
 	}
 
+	// // Dial the Ethereum RPC server.
```
Comment (minhd-vu) [inline cmd/loadtest/loadtest.go:354]: nit double `//`

File: cmd/loadtest/loadtest.go [hunk 2]
```diff
@@ -347,32 +351,48 @@ func runLoadTest(ctx context.Context) error {
...
 		log.Error().Err(err).Msg("Unable to dial rpc")
 		return err
 	}
+	defer rpc.Close()
 	rpc.SetHeader("Accept-Encoding", "identity")
 	ec := ethclient.NewClient(rpc)
 
+	// Define the main loop function.
+	// Make sure to define any logic associated to the load test (initialization, main load test loop
+	// or ven completion steps) in this function in order to handle cancellation signals properly.
```
Comment (minhd-vu) [inline cmd/loadtest/loadtest.go:366]: typo? "when completion stops"?

File: cmd/loadtest/loadtest.go [hunk 3]
```diff
@@ -347,32 +351,48 @@ func runLoadTest(ctx context.Context) error {
...
+			log.Error().Err(err).Msg("Encountered error while wrapping up loadtest")
+		}
+		return nil
 	}
 
+	// Set up signal handling for interrupts.
 	sigCh := make(chan os.Signal, 1)
 	signal.Notify(sigCh, os.Interrupt)
 
+	// // Initialize channels for handling errors and running the main loop.
```
Comment (minhd-vu) [inline cmd/loadtest/loadtest.go:388]: double `//`

File: cmd/loadtest/loadtest.go [hunk 4]
```diff
@@ -347,32 +351,48 @@ func runLoadTest(ctx context.Context) error {
...
 	loadTestResults = make([]loadTestSample, 0)
 	errCh := make(chan error)
 	go func() {
 		errCh <- loopFunc()
 	}()
 
+	// // Wait for the load test to complete, either due to time limit, interrupt signal, or completion.
```
Comment (minhd-vu) [inline cmd/loadtest/loadtest.go:395]: `double //`

File: cmd/loadtest/loadtest.go [hunk 5]
```diff
@@ -347,32 +351,48 @@ func runLoadTest(ctx context.Context) error {
...
-		return mainLoop(ctx, ec, rpc)
+		if mainLoop(ctx, ec, rpc) != nil {
+			log.Error().Err(err).Msg("Error during the main load test loop")
+			return err
+		}
+
+		if completeLoadTest(ctx, ec, rpc) != nil {
+			log.Error().Err(err).Msg("Encountered error while wrapping up loadtest")
+		}
+		return nil
```
Comment (minhd-vu) [inline cmd/loadtest/loadtest.go:381]: why not return complete load test `err`?

File: cmd/loadtest/loadtest.go [hunk 6]
```diff
@@ -347,32 +351,48 @@ func runLoadTest(ctx context.Context) error {
...
 
+	// Define the main loop function.
+	// Make sure to define any logic associated to the load test (initialization, main load test loop
+	// or ven completion steps) in this function in order to handle cancellation signals properly.
 	loopFunc := func() error {
-		err = initializeLoadTestParams(ctx, ec)
-		if err != nil {
+		if initializeLoadTestParams(ctx, ec) != nil {
+			log.Error().Err(err).Msg("Error initializing load test parameters")
 			return err
```
Comment (rebelArtists) [inline cmd/loadtest/loadtest.go:370]: maybe i'm wrong, but where does err var get initialized? is it using an err var from earlier?

File: cmd/loadtest/loadtest.go (hunk 5 above)
Comment (leovct) [reply cmd/loadtest/loadtest.go:381]: because I wanted to display a custom error but I can do this yep

File: cmd/loadtest/loadtest.go (hunk 6 above)
Comment (leovct) [reply cmd/loadtest/loadtest.go:370]: I just forgot to push the [commit](https://github.com/maticnetwork/polygon-cli/pull/147/commits/cf198a93cd9acac605aa8add02e42c2a004ffe23)... 🤦

**Review by minhd-vu — APPROVED** [review]

**Pull request merged**